keps/sig-auth/20190607-certificates-api.md

```diff
@@ -0,0 +1,285 @@
+out-of-band verification of the CSR before making an authorization decision.
+
+The approver is designed to be explicitly independent of the signer. This
+separates concerns of authorization and certificate minting and also signer to
+operate in a separate, more restrictive environment than the approver. This is
+typical of many PKI architectures.
+
+### Sequence of an Issuance
+
+A typical successful issuance proceeds as follows.
+
+![CSR](/keps/sig-auth/csr.png)
+
+1. The requestor generates a private key, builds a certificate signing request,
+   and Submits the certificate request to the Kubernetes certificates API.
+1. The approver controller Observes the newly submitted request, Validates and
+   authorizes the request And if all goes well, approves the request
+1. The signer observes the approval and mints a new certificate.
+1. The requestor downloads it’s certificate.
+
+### CertificateSigningRequest API Definition
+
+```go
+// This information is immutable after the request is created. Only the Request
+// and Usages fields can be set on creation, other fields are derived by
+// Kubernetes and cannot be modified by users.
+type CertificateSigningRequest struct {
+  Spec   CertificateSigningRequestSpec
+  Status CertificateSigningRequestStatus
+}
+
+
+type CertificateSigningRequestSpec struct {
+  // Base64-encoded PKCS#10 CSR data
+  Request []byte
+
+  // usages specifies a set of usage contexts the key will be
+  // valid for.
+  // See: https://tools.ietf.org/html/rfc5280#section-4.2.1.3
+  //      https://tools.ietf.org/html/rfc5280#section-4.2.1.12
+  Usages []KeyUsage
+
+  // Information about the requesting user.
+  // See user.Info interface for details.
+  Username string
+  // UID information about the requesting user.
+  // See user.Info interface for details.
+  UID string
+  // Group information about the requesting user.
+  // See user.Info interface for details.
+  Groups []string
+  // Extra information about the requesting user.
+  // See user.Info interface for details.
+  Extra map[string]string
+}
+
+// ExtraValue masks the value so protobuf can generate
+type ExtraValue []string
+
+type CertificateSigningRequestStatus struct {
+  // Conditions applied to the request, such as approval or denial.
+  Conditions []CertificateSigningRequestCondition
+
+  // If request was approved, the controller will place the issued
+  // certificate here.
+  Certificate []byte
+}
+
+type CertificateSigningRequestCondition struct {
+  // request approval state, currently Approved or Denied.
+  Type RequestConditionType
+  // brief reason for the request state
+  Reason string
+  // human readable message with details about the request state
+  Message string
+}
+
+type RequestConditionType string
+
+// These are the possible conditions for a certificate request.
+const (
+  CertificateApproved RequestConditionType = "Approved"
+  CertificateDenied   RequestConditionType = "Denied"
+)
+```
+
+### Automatic CSR Approval
```

Comment:
"Automatic CSR Approval **implementations**"?
I'm not sure these are worth listing in the KEP